zebra-network/src/peer_set/candidate_set/tests/vectors.rs

```diff
@@ -135,3 +240,49 @@ fn mock_gossiped_peers(last_seen_times: impl IntoIterator<Item = DateTime<Utc>>)
         })
         .collect()
 }
+
+/// Create a mock `PeerSet` service that checks that requests to it are rate limited.
+///
+/// The function also returns an atomic counter, that can be used for checking how many times the
+/// service was called.
+fn mock_peer_service<E>() -> (
+    impl Service<
+            Request,
+            Response = Response,
+            Future = future::Ready<Result<Response, E>>,
+            Error = E,
+        > + 'static,
+    Arc<AtomicUsize>,
+) {
+    let rate_limit_interval = MIN_PEER_GET_ADDR_INTERVAL;
+
+    let call_counter = Arc::new(AtomicUsize::new(0));
+    let call_counter_to_return = call_counter.clone();
+
+    let mut peer_request_tracker: VecDeque<_> =
+        iter::repeat(Instant::now()).take(GET_ADDR_FANOUT).collect();
+
+    let service = tower::service_fn(move |request| {
+        match request {
+            Request::Peers => {
+                // Get time from queue that the request is authorized to be sent
+                let authorized_request_time = peer_request_tracker
+                    .pop_front()
+                    .expect("peer_request_tracker should always have GET_ADDR_FANOUT elements");
+                // Check that the request was rate limited
+                assert!(Instant::now() >= authorized_request_time);
+                // Push a new authorization, updated by the rate limit interval
+                peer_request_tracker.push_back(Instant::now() + rate_limit_interval);
+
+                // Increment count of calls
+                call_counter.fetch_add(1, Ordering::Relaxed);
```

Comment:
Can you use the strongest `SeqCst` memory ordering for all reads and writes to this atomic?

There's a bit more detail about atomics here:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0011-async-rust-in-zebra.md#using-atomics

Also specifically for this use case:
- tokio's watch channel uses `SeqCst` for a similar read/write: https://docs.rs/tokio/1.6.1/src/tokio/sync/watch.rs.html#286
- x86_64 doesn't really implement Relaxed orderings, so we're effectively getting `AcqRel` during testing anyway: https://stackoverflow.com/questions/10537810/memory-ordering-restrictions-on-x86-architecture#18512212

Reply:
Yeah, I had chosen `Relaxed` because in this case there's no data dependencies to that counter. `tokio::sync::watch` uses stricter atomics (AFAICT) because the `Shared` type has _two_ atomics (among other fields), so using `Relaxed` there would mean that the `Shared` type risked having incomplete updates (i.e., one field has a newer value and one has an older value). At least that's according to my current understanding :thinking: 

Using stricter atomics here helps with preventing future issues if a data dependency is added and the ordering requirement is not changed. But since this is a test, I hadn't even considered future-proofing it :sweat_smile:

Reply:
Sometimes it's just easier to use the safe option.